examples/MLIRVector/vector-broadcast.mlir

```diff
@@ -8,15 +8,136 @@ func.func @main() -> i32 {
+  func.call @broadcast_low_dim_to_high_dim(%v1) : (vector<3x1xf32>) -> vector<4x3x2xf32>
+  func.call @broadcast_1_to_n_case(%v2) : (vector<1x3x1xf32>) -> vector<4x3x2xf32>
+  func.call @broadcast_n_to_n_case(%v1) : (vector<3x1xf32>) -> vector<3x2xf32>
+
   %ret = arith.constant 0 : i32
   return %ret : i32
 }
+
+func.func @broadcast_scalar_to_vector(%src: f32) -> vector<3x1xf32> {
+  // broadcast a scalar to vector
+  %result = vector.broadcast %src : f32 to vector<3x1xf32>
+
+  // equals to splat it, like "std::fill" in C++.
+  // %w will be [[%src], [%src], [%src]]
+  %w = vector.splat %src : vector<3x1xf32>
+
+  // %w == %result
+  vector.print %result : vector<3x1xf32>
+  vector.print %w : vector<3x1xf32>
+
+  return %result : vector<3x1xf32>
+}
+
+func.func @broadcast_low_dim_to_high_dim(%src: vector<3x1xf32>) -> vector<4x3x2xf32> {
+  // broadcasting vector wit smaller rank to larger one can be transformed into 
+  // broadcasting two vectors with the same rank
+  %result = vector.broadcast %src : vector<3x1xf32> to vector<4x3x2xf32>
+
+  // We can extend a vector to any higher dimension by adding dimensions with 
+  // length "1" in front of it. For example, extend vector<3x1xf32> to 
+  // vector<1x3x1xf32>, or vector<1x1x1x1x3x1xf32> if we need.
+  %zero = arith.constant 0.0 : f32
+  %t0 = vector.broadcast %zero : f32 to vector<1x3x1xf32>
+  %t1 = vector.insert %src, %t0[0] : vector<3x1xf32> into vector<1x3x1xf32>
+
+  // Then we need to broadcast two rank-equal vectors, which can be done 
+  // recursively on each dimension. Please check @broadcast_1_to_n_cast for details.
+  %t2 = func.call @broadcast_1_to_n_case(%t1) : (vector<1x3x1xf32>) -> vector<4x3x2xf32>
+
+  // %result == %t2
+  vector.print %result : vector<4x3x2xf32>
+  vector.print %t2 : vector<4x3x2xf32>
+
+  return %result : vector<4x3x2xf32>
+}
+
+func.func @broadcast_1_to_n_case(%src: vector<1x3x1xf32>) -> vector<4x3x2xf32> {
+  // Case 1, "1->n" kind. 
+  //    broadcast %src: vector<1 x ... x T> to vector<n x ... x T> 
+  %result = vector.broadcast %src : vector<1x3x1xf32> to vector<4x3x2xf32>
+
+  // let %src == [%e0], then the new vector will be:
+  //    [broadcast %sub : vector<... x T>, ..., broadcast %sub : vector<... x T>]
+
+  // for example, here we get %src = [%e0], n = 4
+  %e0 = vector.extract %src[0] : vector<1x3x1xf32>  
+  
+  // then make %t0, %t1, %t2, %t3 to be broadcast %e0 : vector<... x T>
+  %t0 = vector.broadcast %e0 : vector<3x1xf32> to vector<3x2xf32>
+  %t1 = vector.broadcast %e0 : vector<3x1xf32> to vector<3x2xf32>
+  %t2 = vector.broadcast %e0 : vector<3x1xf32> to vector<3x2xf32>
+  %t3 = vector.broadcast %e0 : vector<3x1xf32> to vector<3x2xf32>
+  
+  // then create vector as [%t0, %t1, %t2, %t3]
+  %zero = arith.constant 0.0 : f32
+  %w_ = vector.broadcast %zero : f32 to vector<4x3x2xf32>
+  %w0 = vector.insert %t0, %w_[0] : vector<3x2xf32> into vector<4x3x2xf32>
+  %w1 = vector.insert %t0, %w0[1] : vector<3x2xf32> into vector<4x3x2xf32>
```

Comment:
Shouldn't this be `t1`? Below ones should be `t2` and `t3` right?

Reply:
My apologies, it should be `t1`, `t2`, ....